compiler/src/iree/compiler/Codegen/Common/test/decompose_softmax.mlir

```diff
@@ -19,25 +19,28 @@ func.func @softmax(%arg0: tensor<2x16x32xf32>) -> tensor<2x16x32xf32> {
 // CHECK:          %[[D8:.+]] = arith.maximumf %[[IN]], %[[OUT]] : f32
 // CHECK:          linalg.yield %[[D8]] : f32
 // CHECK:        } -> tensor<2x16xf32>
-// CHECK:        %[[CST0:.+]] = arith.constant 0.0
-// CHECK:        %[[D4:.+]] = linalg.fill ins(%[[CST0]] : f32) outs(%[[D1]] : tensor<2x16xf32>) -> tensor<2x16xf32>
-// CHECK:        %[[D5:.+]] = linalg.generic {indexing_maps = [#[[MAP]], #[[MAP1]], #[[MAP1]]], iterator_types =
-// CHECK-SAME:     ["parallel", "parallel", "reduction"]} ins(%[[ARG0]], %[[D3]] : tensor<2x16x32xf32>, tensor<2x16xf32>)
-// CHECK-SAME:     outs(%[[D4]] : tensor<2x16xf32>) {
+// CHECK:        %[[D4:.+]] = linalg.generic {indexing_maps = [#[[MAP]], #[[MAP1]], #[[MAP]]], iterator_types = ["parallel",
+// CHECK-SAME:     "parallel", "parallel"]} ins(%[[ARG0]], %3 : tensor<2x16x32xf32>, tensor<2x16xf32>) outs(%[[D0]] : tensor<2x16x32xf32>) {
 // CHECK:        ^bb0(%[[IN:.+]]: f32, %[[IN_1:.+]]: f32, %[[OUT:.+]]: f32):
 // CHECK:          %[[D8:.+]] = arith.subf %[[IN]], %[[IN_1]] : f32
 // CHECK:          %[[D9:.+]] = math.exp %[[D8]] : f32
-// CHECK:          %[[D10:.+]] = arith.addf %[[D9]], %[[OUT]]
-// CHECK:          linalg.yield %[[D10]] : f32
+// CHECK:          linalg.yield %[[D9]] : f32
+// CHECK:        } -> tensor<2x16x32xf32>
+// CHECK:        %[[CST0:.+]] = arith.constant 0.000000e+00 : f32
+// CHECK:        %[[D5:.+]] = linalg.fill ins(%[[CST0]] : f32) outs(%[[D1]] : tensor<2x16xf32>) -> tensor<2x16xf32>
+// CHECK:        %[[D6:.+]] = linalg.generic {indexing_maps = [#[[MAP]], #[[MAP1]]], iterator_types =
+// CHECK-SAME:     ["parallel", "parallel", "reduction"]} ins(%[[D4]] : tensor<2x16x32xf32>)
+// CHECK-SAME:     outs(%[[D5]] : tensor<2x16xf32>) {
+// CHECK:        ^bb0(%[[IN:.+]]: f32, %[[OUT:.+]]: f32):
+// CHECK:          %[[D8:.+]] = arith.addf %[[IN]], %[[OUT]]
+// CHECK:          linalg.yield %[[D8]] : f32
 // CHECK:        } -> tensor<2x16xf32>
-// CHECK:        %[[D7:.+]]:2 = linalg.generic {indexing_maps = [#[[MAP]], #[[MAP1]], #[[MAP1]], #[[MAP]], #[[MAP]]], iterator_types =
-// CHECK-SAME:     ["parallel", "parallel", "parallel"]} ins(%[[ARG0]], %[[D3]], %[[D5]] : tensor<2x16x32xf32>, tensor<2x16xf32>, tensor<2x16xf32>)
-// CHECK-SAME:     outs(%[[D0]], %[[D0]] : tensor<2x16x32xf32>, tensor<2x16x32xf32>) {
-// CHECK:        ^bb0(%[[IN:.+]]: f32, %[[IN_1:.+]]: f32, %[[IN_2:.+]]: f32, %[[OUT0:.+]]: f32, %[[OUT1:.+]]: f32):
-// CHECK:          %[[D8:.+]] = arith.subf %[[IN]], %[[IN_1]] : f32
-// CHECK:          %[[D9:.+]] = math.exp %[[D8]] : f32
-// CHECK:          %[[D10:.+]] = arith.divf %[[D9]], %[[IN_2]] : f32
-// CHECK:          linalg.yield %[[D9]], %[[D10]] : f32
-// CHECK:        } -> (tensor<2x16x32xf32>, tensor<2x16x32xf32>)
-// CHECK:        return %[[D7]]#1 : tensor<2x16x32xf32>
+// CHECK:        %[[D7:.+]] = linalg.generic {indexing_maps = [#[[MAP]], #[[MAP1]], #[[MAP]]], iterator_types =
```

Comment:
> There are few failures with softMax and they are related to missing Vulcan Extension. The code with the changes in this PR causes the pattern matchers to match for VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_COOPERATIVE_MATRIX_FEATURES_KHR extension (thanks @ScottTodd for pointing where the definition for 1000506000!). 
> ```
> exec ${PAGER:-/usr/bin/less} "$0" || exit 1
> Executing tests from //tests/e2e/regression:check_regression_vulkan-spirv_softmax.mlir
> -----------------------------------------------------------------------------
> [==========] Running 2 tests from 1 test suite.
> [----------] Global test environment set-up.
> [----------] 2 tests from module
> [ RUN ] module.softmax
> /tmp/swiftshader_src_U4O4VS/src/Vulkan/VkPhysicalDevice.cpp:649 WARNING: UNSUPPORTED: curExtension->sType: 1000506000
> 
> The error 1000506000 is
> // Provided by VK_KHR_cooperative_matrix
>     VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_COOPERATIVE_MATRIX_FEATURES_KHR = 1000506000,
> ```
> I'm sure there had been similar problems with this kind of errors and I wonder is someone here knows of examples/ of ways how to deal with these.
> For example there should be probably a way to require for a test to ave certain extensions available for it to run. With that said, I think that the test should be able to run even with no extension available too, maybe?
> How are these type of problems handled? Any examples/patterns used in such cases?

We have labels like `vulkan_uses_vk_khr_shader_float16_int8` and `requires-gpu-rdna3` that can be applied to tests. Each CI build can then filter which tests to include/exclude using those labels.

I'm not sure how we have cooperative matrix wired up right now, but I'd expect it to be optional and enabled based on either the known-at-compile-time Vulkan target or the supported-at-runtime Vulkan device. I don't think for a high level op test we should limit to only devices with a certain extension. It's a bit spooky that this PR can lead to that error, and only in one test... maybe the lowerings for softmax (or some of the included ops) down to SPIR-V are brittle?

Reply:
cc @kuhar

Reply:
> I'm not sure how we have cooperative matrix wired up right now, but I'd expect it to be optional and enabled based on either the known-at-compile-time Vulkan target or the supported-at-runtime Vulkan device. I don't think for a high level op test we should limit to only devices with a certain extension

Yes, this should be the case. What's the exact compilation command used by this test?

Reply:
(I just commented here for threading) The test that failed due to requiring cooperative matrix is `//tests/e2e/regression:check_regression_vulkan-spirv_softmax.mlir` in from https://github.com/openxla/iree/blob/main/tests/e2e/regression/CMakeLists.txt. Compile command is probably `iree-compile --iree-input-type=stablehlo --iree-hal-target-backends=vulkan-spirv tests/e2e/regression/softmax.mlir`

Reply:
Yeah this looks like something that should *not* be doing coop matrix.